### PR TITLE
Api drop dep go rootcerts 26136

### LINKED ISSUE #26136: api: drop dependency on go-rootcerts
The [`github.com/hashicorp/go-rootcerts`](https://pkg.go.dev/github.com/hashicorp/go-rootcerts) module doesn't seem necessary anymore with recent Go versions:
* The package was initially created to workaround a *Darwin-only* [bug](https://github.com/golang/go/issues/14514) that has been fixed in Go 4 years ago 
* I filled a question about the relevance of that module (hashicorp/go-rootcerts#21) almost one year ago, and nobody replied. Obviously nobody is maintaining that module.
* Dropping this dependency will also remove a dependency on [`github.com/mitchellh/go-homedir`](https://pkg.go.dev/github.com/mitchellh/go-homedir) (see also hashicorp/go-rootcerts#22 about another way to drop that dependency)

```console
$ cd api
$ go mod graph | grep go-rootcerts          
github.com/hashicorp/vault/api github.com/hashicorp/go-rootcerts@v1.0.2
github.com/hashicorp/go-rootcerts@v1.0.2 github.com/mitchellh/go-homedir@v1.1.0
github.com/hashicorp/vault/api@v1.0.5-0.20200519221902-385fac77e20f github.com/hashicorp/go-rootcerts@v1.0.1
github.com/hashicorp/go-rootcerts@v1.0.1 github.com/mitchellh/go-homedir@v1.1.0
$ go mod graph | grep go-homedir  
github.com/hashicorp/go-rootcerts@v1.0.2 github.com/mitchellh/go-homedir@v1.1.0
github.com/hashicorp/go-rootcerts@v1.0.1 github.com/mitchellh/go-homedir@v1.1.0 
```

### PR DESCRIPTION
**Pull request by dolmen**
### Description
Fixes #26136: drop dependency on module [`github.com/hashicorp/go-rootcerts`](https://github.com/hashicorp/go-rootcerts) by inlining the CACert/CAPath loading logic.

See details on the rationale in #26136.

### CODE REVIEW

**Comment by aslamovamir** [comment]
Hi @dolmen, thanks for opening the PR with us!
Your PR looks good to me and seems to address the underlying issue you opened before, but there are a few things we would love to see before approving the PR! Again, thanks a lot for your contribution!
First, could you please explain why we need to remove the dependency and what are the implications of removing the dependency and of your suggested changes. This could be done partly by creating an [improvement CHANGELOG](https://github.com/hashicorp/vault/blob/main/changelog/README.md) entree, which is required for the type of PR you opened. Secondly, can we please also add unit tests and possibly manual tests to verify your changes do not break anything and covers any potential edge cases. Third, can we please add go docs to each of the new function you created, briefly explaining what it does. Adding go doc as a comment to other parts of your changes, if relevant, could also be added to improve clarity.

File: api/client.go [hunk 1]
```diff
@@ -291,6 +293,56 @@ func DefaultConfig() *Config {
...
 }
 
+func appendCerts(certPool *x509.CertPool, path string) error {
+	pem, err := os.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	if !certPool.AppendCertsFromPEM(pem) {
+		return errors.New("couldn't parse PEM")
```
Comment (aslamovamir) [inline api/client.go:303]:
```suggestion
		return errors.New("could not parse PEM")
```

File: api/client.go [hunk 2]
```diff
@@ -291,6 +293,56 @@ func DefaultConfig() *Config {
...
+}
+
+func loadCACerts(CACertFile string, CACertBytes []byte, CAPath string) (*x509.CertPool, error) {
+	var certPool *x509.CertPool
+
+	switch true {
+	case CACertFile != "":
+		certPool = x509.NewCertPool()
+		if err := appendCerts(certPool, CACertFile); err != nil {
+			return nil, fmt.Errorf("Error loading CA File %s: %w", CACertFile, err)
```
Comment (aslamovamir) [inline api/client.go:315]:
```suggestion
			return nil, fmt.Errorf("Error loading CA File %s: %w", CACertFile, err)
```
```suggestion
			return nil, fmt.Errorf("error loading CA file %s: %w", CACertFile, err)
```

File: api/client.go [hunk 3]
```diff
@@ -291,6 +293,56 @@ func DefaultConfig() *Config {
...
+				return nil
+			}
+
+			if err = appendCerts(certPool, path); err != nil {
+				err = fmt.Errorf("%s: %w", path, err)
+			}
+			return err
+		})
+		if err != nil {
+			return nil, fmt.Errorf("Error loading file from CAPath: %w", err)
```
Comment (aslamovamir) [inline api/client.go:339]:
```suggestion
			return nil, fmt.Errorf("error loading file from CAPath: %w", err)
```

File: api/client.go (hunk 2 above)
Comment (dolmen) [reply api/client.go:315]: This is definitely a change that I wanted to make, but I tried to keep an output similar to go-rootcerts.
https://github.com/hashicorp/go-rootcerts/blob/master/rootcerts.go#L73

This comment just gives more reasons to get rid of go-rootcerts.

File: api/client.go (hunk 3 above)
Comment (dolmen) [reply api/client.go:339]: Like with the other comments, if departing from go-rootcerts output is fine, I'll fix that.

https://github.com/hashicorp/go-rootcerts/blob/b79450d348d0b6b974896f0dd42568742eda45b8/rootcerts.go#L106

File: api/client.go (hunk 1 above)
Comment (dolmen) [reply api/client.go:303]: I took the text from go-rootcerts, and tried to minimize visible changes in error labels. But I'm ok to change as requested.

https://github.com/hashicorp/go-rootcerts/blob/b79450d348d0b6b974896f0dd42568742eda45b8/rootcerts.go#L85

File: api/client.go [hunk 4]
```diff
@@ -291,6 +293,56 @@ func DefaultConfig() *Config {
...
+	switch true {
+	case CACertFile != "":
+		certPool = x509.NewCertPool()
+		if err := appendCerts(certPool, CACertFile); err != nil {
+			return nil, fmt.Errorf("Error loading CA File %s: %w", CACertFile, err)
+		}
+	case len(CACertBytes) > 0:
+		certPool = x509.NewCertPool()
+		if !certPool.AppendCertsFromPEM(CACertBytes) {
+			return nil, errors.New("couldn't parse PEM")
```
[the remark this answers is not in the document]
Comment (dolmen) [reply api/client.go:320]: I'm ok to fix that. The text comes from go-rootcerts.
https://github.com/hashicorp/go-rootcerts/blob/b79450d348d0b6b974896f0dd42568742eda45b8/rootcerts.go#L85

File: api/client.go (hunk 1 above)
Comment (aslamovamir) [reply api/client.go]: Ahh I see, sorry, then we can keep what you already have!

**Comment by aslamovamir** [comment]
@dolmen thanks a lot for addressing the suggestions with error messages: I see that you had a good reason to keep what you had to keep consistency so I am fine with reverting back to your original changes. Can we please also add some unit tests?

**Comment by dolmen** [comment]
@aslamovamir  I had a good reason to keep the labels of the errors, but I don't think that that reason is good enough to revert the errors style which was inconsistent with idiomatic Go.

About unit tests, I'm looking into it, but I found issues #29545 and hashicorp/go-rootcerts#26. So a prerequisite for writing test is to have root certificates valid far in the future, so that's what I'm looking at for now.
